controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,319 @@
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwarecomponents/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwarecomponents/finalizers,verbs=update
+
+// Reconcile handles changes to HostFirmwareComponents resources.
+func (r *HostFirmwareComponentsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	reqLogger := r.Log.WithValues("hostfirmwarecomponents", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3api.BareMetalHost{}
+	err = r.Get(ctx, req.NamespacedName, bmh)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		reqLogger.Info("could not get baremetal host, not running hostfirmwarecomponents reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: resourceNotAvailableRetryDelay}, nil
+	}
+
+	if hasDetachedAnnotation(bmh) {
+		reqLogger.Info("the host is detached, not running hostfirmwarecomponents reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: unmanagedRetryDelay}, nil
+	}
+	// If the reconciliation is paused, requeue
+	annotations := bmh.GetAnnotations()
+
+	if _, ok := annotations[metal3api.PausedAnnotation]; ok {
+		reqLogger.Info("host is paused, no work to do")
+		return ctrl.Result{Requeue: true, RequeueAfter: subResourceNotReadyRetryDelay}, nil
+	}
+
+	// Fetch the HostFirmwareComponents
+	hfc := &metal3api.HostFirmwareComponents{}
+	info := &rhfcInfo{ctx: ctx, log: reqLogger, hfc: hfc, bmh: bmh}
+	if err = r.Get(ctx, req.NamespacedName, hfc); err != nil {
+		// The HFC resource may have been deleted
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Info("hostFirmwareComponnets not found")
+			return ctrl.Result{Requeue: false}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, fmt.Errorf("could not load hostFirmwareComponents: %w", err)
+	}
+
+	// Create a provisioner to access Ironic API
+	prov, err := r.ProvisionerFactory.NewProvisioner(ctx, provisioner.BuildHostDataNoBMC(*bmh), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create provisioner: %w", err)
+	}
+
+	ready, err := prov.TryInit()
+	if err != nil || !ready {
+		var msg string
+		if err == nil {
+			msg = "not ready"
+		} else {
+			msg = err.Error()
+		}
+		reqLogger.Info("provisioner is not ready", "Error", msg, "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
+	}
+
+	newStatus, err := r.updateHostFirmware(info)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not update hostfirmwarecomponents: %w", err)
+	}
+
+	// Check ironic for the components information if possible
+	components, err := prov.GetFirmwareComponents()
+	info.log.Info("retrieving firmware components and saving to resource", "node", bmh.Status.Provisioning.ID)
+
+	if err != nil {
```

Comment:
We ignored this concern with HFS, but I wonder if we need to explicitly handle the situation where a HFC object is created (by a user) for a host with an incompatible driver (think IPMI). Right now, we will enter a loop, which is never going to do anything useful. The problem: we don't communicate anything to the user (who cannot read the logs). I think we should handle this sort of problems explicitly (to distinguish it from "ironic is busy" sort of issues) and use Conditions to say "This HFC is invalid".

Reply:
Humm, yeah if the user creates and the BMH has an incompatible driver the call 
```
components, err := prov.GetFirmwareComponents()
``` 
will return an error. I will update to not requeue here and see how to use conditions in this part. 
Thanks!

Reply:
Since I added a new error directly to the `GetFirmwareComponents` I don't think I need to check things here again.

Reply:
I don't think the conditions part has been addressed? The logs are not user-visible.

Can be addressed in a follow-up if there are no other critical comments. Let's just not forget about it.

Reply:
Oh , I totally missed the part about conditions, basically are you saying we should add 

```
setUpdatesCondition(generation , status , info, cond, newStatus, reason, message) 
```


and `message` would be something along the lines  "This HFC is invalid" ?